plugins/hive-plugin/src/main/java/edu/berkeley/ground/plugins/hive/GroundReadWriteNodeResource.java

```
@@ -0,0 +1,116 @@
+        try {
+            String response = PluginUtil.execute(get);
+            if (response != null) {
+                return this.constructNode(response);
+            }
+            return null;
+        } catch (IOException e) {
+            throw new GroundException(e);
+        }
+    }
+
+    public List<Long> getTransitiveClosure(Long nodeVersionId) throws GroundException {
+        GetMethod get = new GetMethod(PluginUtil.groundServerAddress + "nodes/closure/" + nodeVersionId);
+        try {
+            return (List<Long>) PluginUtil.getVersionList(get);
+        } catch (IOException e) {
+            throw new GroundException(e);
+        }
+    }
+
+    public List<Long> getAdjacentNodes(Long prevVersionId, String edgeName) throws GroundException {
+        GetMethod get = new GetMethod(PluginUtil.groundServerAddress + "adjacent/" + prevVersionId + "/" + edgeName);
+        try {
+            return (List<Long>) PluginUtil.getVersionList(get);
+        } catch (IOException e) {
+            throw new GroundException(e);
+        }
+    }
+
+    // create a node using Tag
+    Node createNode(String name, Map<String, Tag> tagMap) throws GroundException {
+        try {
+            String encodedUri = PluginUtil.groundServerAddress + "nodes/" + URLEncoder.encode(name, "UTF-8");
+            PostMethod post = new PostMethod(encodedUri);
+            ObjectMapper mapper = new ObjectMapper();
+            String jsonString = mapper.writeValueAsString(tagMap);
+            post.setRequestEntity(PluginUtil.createRequestEntity(jsonString));
+            String response = PluginUtil.execute(post);
+            return this.constructNode(response);
+        } catch (IOException e) {
+            throw new GroundException(e);
+        }
+
+    }
+
+    // helper methods for Node and NodeVersion
+    private Node constructNode(String response) throws GroundException {
+        if (response == null) {
```

(I've a personal aversion to handling with `null`s)

---

In general, what do you suggest for this? I was originally using `Optional`s for this as is, from my understanding, canonical in Scala, but upon further research / reading, it seems like the creators of the Java `Optional` vehemently oppose that sort of usage.

---

I guess that depends on the semantics of the method. In this case, we want to construct something from something, so in the case that construction is impossible, I would throw a checked exception. Probably a subclass of `GroundException`. (That would be in Java... as you mention, in Scala, I would probably use an `Option` or a `Try` if I want to preserve the error cause)